### DISCUSSION TITLE
Next 13 passing data between server components

### DISCUSSION THREAD

**Discussion by learnwithsumit** [vercel/next.js#42861] [Help]
As we know, **Next 13** encourages data fetching at **server components** as much as possible. We can get data at the layout level for each route segment and thereby resolve data waterfalls. But what about **authentication**? Suppose, I do a fetch request at _layout.js_ file inside app directory (which is a server component) and get jwt access token from my API. Now how can I pass that access token to other server components deeply nested inside? Because I need data fetching inside nested server components as well and in that case I need to pass the access token! Can anyone help?

**Comment by pedrosandes** [comment]
Maybe you can use headers()

https://beta.nextjs.org/docs/api-reference/headers

**Comment by sumitsaha** [comment]
But based on the documentation, the headers are read only. We can't set anything there. Suppose, in the app/layout.js file (which is a server component), I send a request to my auth api endpoint and get access token. Then I want to pass that access token deep down the component tree. How a server component in the tree can use that access token for further data fetching in protected routes?

**Reply by VikashLoomba** [reply]
The problem of making an ephemeral token needed for API requests available to server components has been my issue too. Have you found any solutions? It seems that being able to write to the headers object is the obvious solution, but I haven't found any solutions.

Here is my post describing something similar: https://github.com/vercel/next.js/discussions/43583

**Reply by learnwithsumit** [reply]
Thanks @VikashLoomba ! I also didn't find any proper solution yet! I tried with Next Auth and it saves token is session cookie which I can access by requesting the "/api/auth/session" endpoint from inside my server component like below:

```js
export default async function getSession(cookie) {
    const response = await fetch("http://localhost:3000/api/auth/session", {
        headers: {
            cookie,
        },
    });

    const session = await response.json();

    return Object.keys(session).length > 0 ? session : null;
}
```
